### PR TITLE
Fix #8645: no assertion warning on special modes (help, version, etc.)

### LINKED ISSUE #8645: Warning: "debug build with assertions" interferes with --numeric-version
The warning printed by the `cabal` development version interferes with collecting output, e.g.
```sh
CABAL_VERSION=$(cabal --numeric-version)
```
will usually set `CABAL_VERSION` to something like `3.8.1.0`, but for the development version it is the warning-string followed by the version:
```console
$ echo $CABAL_VERSION
Warning: this is a debug build of cabal-install with assertions enabled. 3.9
```

The problem is that this warning is printed to `stdout` with `normal` verbosity:
https://github.com/haskell/cabal/blob/399cf6432df746b0b0d8e33a8b41736120552380/cabal-install/main/Main.hs#L192-L198

I think this warning should only be printed with increased verbosity, e.g. `-v2`.

What confuses me is that `-v0` does not seem to have any effect:
```
$ cabal -v0 --numeric-version
Warning: this is a debug build of cabal-install with assertions enabled.
3.9
```
I still see this well-meant but silly warning.

### LINKED ISSUE #8654: cabal gen-bounds: Consume (multiple) build plans?
I’d like to automate the managing of my build dependencies more, and `cabal gen-bounds` sounds like it should be an important tool in the process.

However, my projects usually support a range of builds (e.g. different GHC versions), and simply generating the bounds based on _one_ solver results is not going to work.

I could imagine that `cabal gen-bounds` could read in multiple build plans (e.g. the `plan.json` produced when building the project) and generating the appropriate dependencies (maybe in `^>= { 1.2.3, 3.4.5 }` form) from the union of these build plans.

(Also see related discussion in https://discourse.haskell.org/t/dependency-version-bounds-are-a-lie/5522/17?u=nomeata)

### PR DESCRIPTION
**Pull request by andreasabel**
Fix #8645: no assertion warning on special modes (help, version, etc.); print warning to `stderr`.

This PR moves the `warnIfAssertionsAreEnabled` until after the main command line argument parsing, and then only prints it on regular operation, i.e. not when just the help or version of cabal was demanded, or the command line could not be parsed.

Further, the warning is printed on stderr.  The previous attempt to use `warn` does not work as expected, as the verbosity aperatshnik isn't in place yet, so the warning cannot be suppressed by -v0.

Drive-by-shooting: straighten the convoluted control flow involving `maybeScriptAndArgs`.  This IO computation was invoked "brutto" before, when it is actually only needed in *one* branch of the case distinction.

### CODE REVIEW

**Review by Mikolaj — APPROVED** [review]
A nitpick: the refactoring and the functionality change would ideally be in separate commits.

**Comment by SaadAhmedGit** [comment]
Shouldn't the warning message be:

> "**Warning:** this is a debug build of cabal-install with assertions enabled."

instead of
> "this is a debug build of cabal-install with assertions enabled."

Since the string "Warning: " was being prepended to the message automatically with `warn` which has now been removed thus, requiring the word to be written explicitly in the message. At least, it was that way when it printed to `stdout` before using `warn`.

**Comment by ulysses4ever** [comment]
If the reason for doing away with `warn` is that -v0 is not parsed yet, maybe `warnIfAssertionsAreEnabled` could be pushed even further, to the point where -v0 is parsed and the verbosity machinery is up?

File: cabal-install/main/Main.hs [hunk 1]
```diff
@@ -189,20 +188,20 @@ main = do
...
 warnIfAssertionsAreEnabled =
   assert False (return ()) `catch`
-  (\(_e :: AssertionFailed) -> warn normal assertionsEnabledMsg)
+  (\(_e :: AssertionFailed) -> hPutStrLn stderr assertionsEnabledMsg)
+    -- Andreas, 2022-12-30, issue #8654:
+    -- The verbosity machinery is not in place at this point (option -v not parsed),
+    -- so instead of using function @warn@, we print straight to stderr.
   where
     assertionsEnabledMsg =
       "this is a debug build of cabal-install with assertions enabled."
```
Comment (SaadAhmedGit) [inline cabal-install/main/Main.hs:201]: prepend "Warning: " to the message

Comment (SaadAhmedGit) [inline cabal-install/main/Main.hs:201]: Prepend "Warning: " to the message.

**Review by SaadAhmedGit — CHANGES_REQUESTED** [review]

**Comment by jneira** [comment]
Thanks for the fix, it makes sense to not warn on those commands. Some chance of having a test for this?

**Comment by Mikolaj** [comment]
Ouch, from all this enthusiasm, I forgot to ask for a test and a changelog file.

**Comment by andreasabel** [comment]
> Ouch, from all this enthusiasm, I forgot to ask for a test and a changelog file.

Test: It seems we didn't have a test for this in the first place, otherwise it would have been broken by this PR.  What kind of test do you imagine for this?  Does CI run on tests on a `cabal-install` with assertions on?  Which test would be a model for the test envisioned here?

Changelog: I think this should not go into the changelog, as it does not affect regular users of `Cabal` and `cabal-install`.  If I am not mistaken, assertions will be off when building from Hackage...

**Comment by Mikolaj** [comment]
> Test: It seems we didn't have a test for this in the first place, otherwise it would have been broken by this PR.

Indeed, that's strange.

> What kind of test do you imagine for this? Does CI run on tests on a `cabal-install` with assertions on?

CI definitely should run only with assertions on. But perhaps it calls cabal-install commands via some internal API bypassing the warning printout? I think we should have at least one test where the `cabal-install` binary is called directly (on one of the commands that trigger the warning)/

> Which test would be a model for the test envisioned here?

There may not be one.

> Changelog: I think this should not go into the changelog, as it does not affect regular users of `Cabal` and `cabal-install`. If I am not mistaken, assertions will be off when building from Hackage...

Yes, assertions are off on Hackage. This is debatable (changelog is useful for all consumers of cabal, including devs and bug reporters, the latter of which get confused most easily by the failed asssertions when reproducing with master branch artifacts from GHA), but I won't insist.

File: cabal-install/main/Main.hs (hunk 1 above)
Comment (andreasabel) [reply cabal-install/main/Main.hs]: Good, catch @SaadAhmedGit , fixed this.

**Comment by andreasabel** [comment]
@ulysses4ever :
> If the reason for doing away with `warn` is that -v0 is not parsed yet, maybe `warnIfAssertionsAreEnabled` could be pushed even further, to the point where -v0 is parsed and the verbosity machinery is up?

It seems pushing this further in would require to prepend it to _every_ command.  I only pushed it as far in so that it is still before we branch into the individual commands.

**Comment by andreasabel** [comment]
@Mikolaj: Ok, I added a changelog.

PR needs **SQUASHING** to integrate this into the main commit.

**Comment by Mikolaj** [comment]
> @Mikolaj: Ok, I added a changelog.

Thank you.

> PR needs **SQUASHING** to integrate this into the main commit.

I think, if you set the label already, it will get squashed at the right time (meaning, after reviews and 2 days).

**Review by SaadAhmedGit — COMMENTED** [review]
LGTM

**Review by SaadAhmedGit — APPROVED** [review]

**Comment by ulysses4ever** [comment]
@andreasabel 

> It seems pushing this further in would require to prepend it to every command.

But it means, then, that `-v0` is handled in every command? I hoped it's not the case. If it is, maybe a refactoring pulling it up would make sense, and that would allow to push the assertion message after it still?

**Comment by andreasabel** [comment]
> @ulysses4ever wrote:
> But it means, then, that `-v0` is handled in every command? I hoped it's not the case. If it is, maybe a refactoring pulling it up would make sense, and that would allow to push the assertion message after it still?

Artem, please have a look at https://github.com/haskell/cabal/blob/7babda92df9e005185e7c9d5e9308a8a96092d7c/cabal-install/main/Main.hs and observe the amount of code duplication there.  I think refactoring this is a noble goal, but too much to require for this modest PR.

**Comment by ulysses4ever** [comment]
@andreasabel fair enough!

**Review by ulysses4ever — APPROVED** [review]

**Comment by jneira** [comment]
> Test: It seems we didn't have a test for this in the first place, otherwise it would have been broken by this PR.

That is just the core of the problem (a general problem in the cabal codebase) and no a valid argument to no add them imho
But to be fair if the tested cabal executable has no assertions, it is difficult to add them

**Comment by Mikolaj** [comment]
> if the tested cabal executable has no assertions, it is difficult to add them

I think we might have had this discussion long ago, but I don't remember the conclusions. Naively, I'd think we should only ever test binaries with assertions. Perhaps one odd test without assertions to make sure no functional code got inserted into an assertion by mistake.

And indeed a quick look at `cabal.project.validate` turns up `-fno-ignore-asserts`. What's your evidence we test without assertions?

**Pull request merged**